data-pipeline-ffi/src/trace_exporter.rs

```diff
@@ -1,59 +1,101 @@
 // Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
 // SPDX-License-Identifier: Apache-2.0
 
-use data_pipeline::trace_exporter::TraceExporter;
-use data_pipeline::trace_exporter::TraceExporterBuilder;
+use data_pipeline::trace_exporter::{
+    ResponseCallback, TraceExporter, TraceExporterInputFormat, TraceExporterOutputFormat,
+};
 use ddcommon_ffi::{
     slice::{AsBytes, ByteSlice},
-    CharSlice,
+    CharSlice, MaybeError,
 };
-use std::ffi::c_char;
+use std::{ffi::c_char, ptr::NonNull};
 
+/// Create a new TraceExporter instance.
+///
+/// # Arguments
+///
+/// * `out_handle` - The handle to write the TraceExporter instance in.
+/// * `url` - The URL of the Datadog Agent to communicate with.
+/// * `tracer_version` - The version of the client library.
+/// * `language` - The language of the client library.
+/// * `language_version` - The version of the language of the client library.
+/// * `language_interpreter` - The interpreter of the language of the client library.
+/// * `input_format` - The input format of the traces. Setting this to Proxy will send the trace
+///   data to the Datadog Agent as is.
+/// * `output_format` - The output format of the traces to send to the Datadog Agent. If using the
+///   Proxy input format, this should be set to format if the trace data that will be passed through
+///   as is.
+/// * `agent_response_callback` - The callback into the client library that the TraceExporter uses
+///   for updated Agent JSON responses.
 #[no_mangle]
 pub unsafe extern "C" fn ddog_trace_exporter_new(
-    host: CharSlice,
-    port: u16,
+    out_handle: NonNull<Box<TraceExporter>>,
```

Comment:
This is fine. `NonNull<T>` is a wrapper around a `*mut T`, and memory slots pointed to by raw pointers to don't have to be initialized just not nul (as long as you don't try to dereference or read the pointer) . So having a `NonNull<MaybeUninit<T>>` is semantically equivalent and it's actually safe to cast `NonNull<MaybeUninit<T>>` to `NonNull<T>` because `NonNull` doesn't guarantee it's content.

So `out_handle.ptr().write(Box::new(..))` is safe, but `*out_handle.ptr() = Box::new(..)` is not because it's going to try to run a destructor on the uninitialized box in `NonNull`.

And to be fair, if you think `NonNull<MaybeUninit<Box<TraceExporter>>>` is superior to `NonNull<Box<TraceExporter>>`, I'd argue that `&mut MaybeUninit<Box<TraceExporter>>` is even better in this case.

I wrote the code this way because although the C generated headers are the same, it's nicer to manipulate the functions from rust to write tests [like here](https://github.com/DataDog/libdatadog/blob/main/ddtelemetry-ffi/src/lib.rs#L275).

Reply:
To be clear, I'm not saying any of this is not safe. I'm just saying I sometimes see various of stuff like like `* mut MaybeUninit<T>` and `NonNull<T>` when `T` may be not be initialized. I don't see cohesion in the Rust project when working with these sort of uninitialized types.